### PR TITLE
[Fix #11137] Fix a false positive for `Style/RedundantEach` when using `each` with a symbol proc argument

### LINKED ISSUE #11137: False positive in Style/RedundantEach with block-pass arguments
Given the code:
```
records.each(&:lock!).each do |record|
end
```

`Style/RedundantEach` says "Remove redundant each" on the first `each`.

Note that this passes the cop:
```
records.each { |record| record.lock! }.each do |record|
end
```
So it's only when there is a block-pass that it ignores that there is an actual block implementation.

--------

## Expected behavior

No offences

## Actual behavior

```
test.rb:2:8: C: [Corrected] Style/RedundantEach: Remove redundant each.
records.each(&:lock!).each do |record|
       ^^^^^
```

It's gets corrected to:
```
records(&:lock!).each do |record|
end
```

## RuboCop version
```
rubocop -V
1.38.0 (using Parser 3.1.2.1, rubocop-ast 1.23.0, running on ruby 3.1.2) [arm64-darwin21]
```

### PR DESCRIPTION
**Pull request by ydah**
Fix: #11137

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: changelog/fix_fix_a_false_positive_for_style_redundant_each.md [hunk 1]
```diff
@@ -0,0 +1 @@
+* [#11137](https://github.com/rubocop/rubocop/issues/11137): Fix a false positive for `Style/RedundantEach` with block-pass arguments. ([@ydah][])
```
Comment (koic) [inline changelog/fix_fix_a_false_positive_for_style_redundant_each.md:1]: Can you tweak it?

```suggestion
* [#11137](https://github.com/rubocop/rubocop/issues/11137): Fix a false positive for `Style/RedundantEach` when using `each` with a symbol proc argument. ([@ydah][])
```

File: spec/rubocop/cop/style/redundant_each_spec.rb [hunk 2]
```diff
@@ -181,4 +181,11 @@
       array.each_foo { |i| foo(i) }.each_with_object([]) { |i| bar(i) }
     RUBY
   end
+
+  it 'does not register an offense with block pass arguments' do
```
Comment (koic) [inline spec/rubocop/cop/style/redundant_each_spec.rb:185]:
```suggestion
  it 'does not register an offense when `each` with a symbol proc argument' do
```

File: changelog/fix_fix_a_false_positive_for_style_redundant_each.md (hunk 1 above)
Comment (ydah) [reply changelog/fix_fix_a_false_positive_for_style_redundant_each.md]: Thank you so much. I updated it.

**Comment by ydah** [comment]
I updated this PR. Thank you for your review.

File: lib/rubocop/cop/style/redundant_each.rb [hunk 3]
```diff
@@ -61,6 +61,8 @@ def on_send(node)
 
         # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
         def redundant_each_method(node)
+          return if node.first_argument&.block_pass_type?
```
[the remark this answers is not in the document]
Comment (Darhazer) [reply lib/rubocop/cop/style/redundant_each.rb]: It means that some of the existing test cases may fail if the block is changed for block-pass. If nothing fails, either there are missing test cases, or accidental complexity in the cop

[the remark this answers is not in the document]
Comment (Darhazer) [reply lib/rubocop/cop/style/redundant_each.rb]: OK, I tested which example is affected by the `next if descendant.parent.block_type? ` and found that now:
`array.reverse_each(&:lock!).each { |i| bar(i) }`
will result in a false positive.
It's being fixed by
```
index fa52b34ed..55f9aa13d 100644
--- a/lib/rubocop/cop/style/redundant_each.rb
+++ b/lib/rubocop/cop/style/redundant_each.rb
@@ -68,7 +70,7 @@ module RuboCop
           end
 
           ancestor_node || node.each_descendant(:send).detect do |descendant|
-            next if descendant.parent.block_type?
+            next if descendant.parent.block_type? || descendant.last_argument&.block_pass_type?
```
https://github.com/rubocop/rubocop/pull/11138/files

[the remark this answers is not in the document]
Comment (ydah) [reply lib/rubocop/cop/style/redundant_each.rb]: Sorry for the late reply. Thank you very much. I updated this PR.

File: spec/rubocop/cop/style/redundant_each_spec.rb [hunk 4]
```diff
@@ -181,4 +192,24 @@
...
+  end
+
+  it 'does not register an offense when using `each` with a symbol proc for last argument' do
+    expect_no_offenses(<<~RUBY)
+      array.each(foo, &:bar).each do |i|
+      end
+    RUBY
+  end
+
+  it 'does not register an offense when using `each(&:foo).each {...}`' do
```
Comment (Darhazer) [inline spec/rubocop/cop/style/redundant_each_spec.rb:210]:
```suggestion
  it 'does not register an offense when using `reverse_each(&:foo).each {...}`' do
```

**Review by Darhazer — APPROVED** [review]
Thank you

File: spec/rubocop/cop/style/redundant_each_spec.rb (hunk 4 above)
Comment (ydah) [reply spec/rubocop/cop/style/redundant_each_spec.rb]: Thank you very much. I updated this PR.

**Comment by koic** [comment]
Thanks!

**Pull request merged**